exporter/stackdriver/stackdriver.go

```diff
@@ -0,0 +1,145 @@
+// https://developers.google.com/identity/protocols/application-default-credentials
+//
+// Exporter can be used to propagate traces over HTTP requests for Stackdriver Trace.
+//
+// Example:
+//
+//	import "go.opencensus.io/plugin/http/httptrace"
+//
+//	client := &http.Client {
+//		Transport: httptrace.NewTransport(nil, exporter),
+//	}
+//
+// All outgoing requests from client will include a Stackdriver Trace header.
+// See the httptrace package for how to handle incoming requests.
+package stackdriver
+
+import (
+	"net/http"
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/trace"
+	"google.golang.org/api/option"
+	monitoredrespb "google.golang.org/genproto/googleapis/api/monitoredres"
+)
+
+// Options contains options for configuring the exporter.
+type Options struct {
+	// ProjectID is the identifier of the Stackdriver
+	// project the user is uploading the stats data to.
+	ProjectID string
+
+	// OnError is the hook to be called when there is
+	// an error occurred when uploading the stats data.
+	// If no custom hook is set, errors are logged.
+	// Optional.
+	OnError func(err error)
+
+	// ClientOptions are additional options to be passed
+	// to the underlying Stackdriver Monitoring API client.
+	// Optional.
+	ClientOptions []option.ClientOption
+
+	// BundleDelayThreshold determines the max amount of time
+	// the exporter can wait before uploading view data to
+	// the backend.
+	// Optional.
+	BundleDelayThreshold time.Duration
+
+	// BundleCountThreshold determines how many view data events
+	// can be buffered before batch uploading them to the backend.
+	// Optional.
+	BundleCountThreshold int
+
+	// Resource is an optional field that represents the Stackdriver
+	// MonitoredResource, a resource that can be used for monitoring.
+	// If no custom ResourceDescriptor is set, a default MonitoredResource
+	// with type global and no resource labels will be used.
+	// Optional.
+	Resource *monitoredrespb.MonitoredResource
+}
+
+// Exporter is a stats.Exporter, trace.Exporter and propagation.HTTPFormat
+// implementation that uploads data to Stackdriver.
+type Exporter struct {
+	traceExporter *traceExporter
+	statsExporter *statsExporter
+}
+
+func NewExporter(o Options) (*Exporter, error) {
+	se, err := newStatsExporter(o)
+	if err != nil {
+		return nil, err
+	}
+	te, err := newTraceExporter(o)
+	if err != nil {
+		return nil, err
+	}
+	return &Exporter{
+		statsExporter: se,
+		traceExporter: te,
+	}, nil
+}
+
+// ExportView exports to the Stackdriver Monitoring if view data
+// has one or more rows.
+func (e *Exporter) ExportView(vd *stats.ViewData) {
+	e.statsExporter.ExportView(vd)
+}
+
+// ExportSpan exports a SpanData to Stackdriver Trace.
+func (e *Exporter) ExportSpan(sd *trace.SpanData) {
+	e.traceExporter.ExportSpan(sd)
+}
+
+// ToRequest modifies the given request to include a Stackdriver Trace header.
+func (e *Exporter) ToRequest(sc trace.SpanContext, req *http.Request) {
```

Comment:
I don't think that propagation should be conflated with exporting. It's valid to use just the stackdriver propagation format but not the stackdriver exporters, for example if you're communicating with Google APIs from AWS and using Jaeger as your tracing backend.

Reply:
I will send a separate CL to address this split.